### ISSUE TITLE
SaveChanges does not commit transaction if only commands are executed with session

### ISSUE THREAD

**Issue by VilleHakli** [JasperFx/marten#2137]
If session is only used to execute commands (e.g. `session.ExecuteAsync(command)`) the transaction is not committed during `session.SaveChangesAsync()`. SaveChanges methods seems to check the work tracker for changes and in case only commands are executed, the tracker does not have any changes.

Failing test case:
```csharp
using System;
using System.Threading.Tasks;
using Bug2137;
using Marten;
using Marten.Testing.Harness;
using Npgsql;
using Shouldly;
using Weasel.Core;
using Xunit;

namespace DocumentDbTests.Bugs
{
    public class Bug_2137_save_changes_should_commit_if_only_execute_is_called: BugIntegrationContext
    {
        [Fact]
        public async Task should_commit_transaction_if_only_commands_are_executed_in_session()
        {
            using var documentStore = SeparateStore(x =>
            {
                x.AutoCreateSchemaObjects = AutoCreate.All;
                x.Schema.For<TestEntity>();
            });

            await documentStore.Advanced.Clean.DeleteAllDocumentsAsync();

            await using (var session = documentStore.OpenSession())
            {
                session.Store(new TestEntity { Name = "Test" });
                await session.SaveChangesAsync();
            }

            await using (var session = documentStore.OpenSession())
            {
                await using var command = new NpgsqlCommand("DELETE FROM bugs.mt_doc_testentity");
                await session.ExecuteAsync(command);
                await session.SaveChangesAsync();
            }

            await using (var session = documentStore.OpenSession())
            {
                var entities = await session.Query<TestEntity>().ToListAsync();
                entities.ShouldBeEmpty();
            }
        }
    }
}

namespace Bug2137
{
    public class TestEntity
    {
        public Guid Id { get; set; }
        public string Name { get; set; }
    }
}
```

**Comment by jeremydmiller** [comment]
@VilleHakli The first question is "why would you do this?"

**Comment by VilleHakli** [comment]
We have two cases where this might happen.
1. We have database migrator, which mainly migrates documents & Marten schema changes. The same migrator is also used to run schema changes to "plain old SQL tables" as this way we do not need to have two separate migration solution. This also allows us to have schema & data changes in same transaction if need be.
2. We need to delete all data in the database for single tenant.

Both of these cases are something that we can handle without Marten, so it this is something that is not in scope of Marten, we can roll our own connection handling.

All solutions are fine by me except for the third one. I just couldn't get that to work :D I tried the test case above with both of these, but it still fails.
```csharp
await using (var session = documentStore.OpenSession())
{
    await using var command = new NpgsqlCommand("DELETE FROM bugs.mt_doc_testentity", session.Connection!);
    await command.ExecuteNonQueryAsync();
    await session.SaveChangesAsync();
}

await using (var session = documentStore.OpenSession())
{
    await using var command = session.Connection!.CreateCommand("DELETE FROM bugs.mt_doc_testentity");
    await command.ExecuteNonQueryAsync();
    await session.SaveChangesAsync();
}
```
In case where Marten handles transactions, I think it would be fine to commit open transaction during SaveChanges, even if there are no changes according to work tracker. At least to me it feels weird that if user calls SaveChanges it will end up rolling back the transaction at some point.

There might also be other problem with the SaveChanges as it does not rollback open transaction in case there is "nothing" to do. I'm not sure if this is big problem, but it might cause some weird cases if same session is used with multiple calls to save changes.

**Comment by elexisvenator** [comment]
> There might also be other problem with the SaveChanges as it does not rollback open transaction in case there is "nothing" to do and thus the session will still have open transaction. I'm not sure if this is big problem, but it might cause some weird cases if same session is used with multiple calls to save changes.

This is a major (separate?) issue as well. We are having connection pool drained issues which appear to be caused hundreds of connections in the "idle in transaction" state in postgres, the last command any of the connections did was a simple query.

**Comment by jeremydmiller** [comment]
@elexisvenator That's not related whatever that is. The connection would be closed and disposed in the Marten sessions when they are disposed. First thing I'd tell you is to check your code to make sure things are being properly disposed.

**Comment by VilleHakli** [comment]
There are fixes and changes in version 5.4.0 which changes Marten to not open transaction so eagerly. Those changes also fixes this or at least makes this irrelevant. So closing this.
